pkg/sql/rollback.go

```diff
@@ -0,0 +1,226 @@
+	distSender *kvcoord.DistSender,
+	table catalog.TableDescriptor,
+	predicates roachpb.DeleteRangePredicates,
+	batchSize int64,
+) error {
+
+	log.Infof(ctx, "deleting data for table %d with predicate %s", table.GetID(), predicates.String())
+	tableKey := roachpb.RKey(codec.TablePrefix(uint32(table.GetID())))
+	tableSpan := roachpb.RSpan{Key: tableKey, EndKey: tableKey.PrefixEnd()}
+
+	// TODO (msbutler): tune these
+	rangesPerBatch := int64(100)
+	numworkers := 4
+
+	initialSpans := make(roachpb.Spans, 1)
+
+	// Partition the inputted span such that each span
+	// partition contains 100 ranges of data.
+	var n int64
+	lastKey := tableSpan.Key
+	ri := kvcoord.MakeRangeIterator(distSender)
+	for ri.Seek(ctx, tableSpan.Key, kvcoord.Ascending); ; ri.Next(ctx) {
+		if !ri.Valid() {
+			return ri.Error()
+		}
+		if n++; n >= rangesPerBatch || !ri.NeedAnother(tableSpan) {
+			endKey := ri.Desc().EndKey
+			if tableSpan.EndKey.Less(endKey) {
+				endKey = tableSpan.EndKey
+			}
+			initialSpans = append(initialSpans, roachpb.Span{Key: lastKey.AsRawKey(),
+				EndKey: endKey.AsRawKey()})
+			n = 0
+			lastKey = endKey
+		}
+
+		if !ri.NeedAnother(tableSpan) {
+			break
+		}
+	}
+
+	spansToDo := make(chan *roachpb.Span, len(initialSpans))
+	for _, span := range initialSpans {
+		spansToDo <- &span
+	}
+
+	// Spin up a few workers to issue DeleteRange requests on the spans in the
+	// spansToDo channel. Using several workers allows the distsender to handle
+	// the span partitions of the initial span concurrently.
+	grp := ctxgroup.WithContext(ctx)
+	grp.GoCtx(func(ctx context.Context) error {
+		return ctxgroup.GroupWorkers(ctx, numworkers, func(ctx context.Context, _ int) error {
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case span := <-spansToDo:
+					admissionHeader := roachpb.AdmissionHeader{
+						Priority:                 int32(admissionpb.BulkNormalPri),
+						CreateTime:               timeutil.Now().UnixNano(),
+						Source:                   roachpb.AdmissionHeader_FROM_SQL,
+						NoMemoryReservedAtSource: true,
+					}
+					delRangeRequest := &roachpb.DeleteRangeRequest{
+						RequestHeader: roachpb.RequestHeader{
+							Key:    span.Key,
+							EndKey: span.EndKey,
+						},
+						UseRangeTombstone: true,
+						Predicates:        predicates,
+					}
+					log.VEventf(ctx, 2, "deleting range %s - %s", span.Key, span.EndKey)
+
+					rawResp, err := kv.SendWrappedWithAdmission(
+						ctx,
+						db.NonTransactionalSender(),
+						roachpb.Header{MaxSpanRequestKeys: batchSize},
+						admissionHeader,
+						delRangeRequest)
+
+					if err != nil {
+						return errors.Wrapf(err.GoError(), "delete range %s - %s", span.Key, span.EndKey)
+					}
+					resp := rawResp.(*roachpb.DeleteRangeResponse)
+					if resp.ResumeSpan != nil {
+						if !resp.ResumeSpan.Valid() {
+							return errors.Errorf("invalid resume span: %s", resp.ResumeSpan)
+						}
+						spansToDo <- resp.ResumeSpan
```

Comment:
Oh, I see why this gets tricky now. Would it be better to use a for loop for this rather than feeding it back into the channel? Having the same goroutine both consume and produce messages on a channel is a bit of an antipattern (it muddies the waters wrt. the termination condition, because it isn't clear whose responsibility it is to close the channel).

Reply:
thanks for the wisdom!